8324678: Replace NULL with nullptr in HotSpot gtests

coleenp opened the pull request:
If this is sufficient, here's the change for NULL to nullptr, adjusting some obvious strings that had NULL in them maybe not all.
Tested with gtest,

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8324678](https://bugs.openjdk.org/browse/JDK-8324678): Replace NULL with nullptr in HotSpot gtests (**Enhancement** - P4)


### Reviewers
 * [Kim Barrett](https://openjdk.org/census#kbarrett) (@kimbarrett - **Reviewer**) ⚠️ Review applies to [d8e3d92d](https://git.openjdk.org/jdk/pull/17577/files/d8e3d92d2198cb9c92d1da923f5d1b2772edbd51)
 * [David Holmes](https://openjdk.org/census#dholmes) (@dholmes-ora - **Reviewer**) ⚠️ Review applies to [d8e3d92d](https://git.openjdk.org/jdk/pull/17577/files/d8e3d92d2198cb9c92d1da923f5d1b2772edbd51)
 * [Julian Waters](https://openjdk.org/census#jwaters) (@TheShermanTanker - Committer)


### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/17577/head:pull/17577` \
`$ git checkout pull/17577`

Update a local copy of the PR: \
`$ git checkout pull/17577` \
`$ git pull https://git.openjdk.org/jdk.git pull/17577/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 17577`

View PR using the GUI difftool: \
`$ git pr show -t 17577`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/17577.diff">https://git.openjdk.org/jdk/pull/17577.diff</a>

</details>


### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/17577#issuecomment-1912634434)

kimbarrett commented on test/hotspot/gtest/gc/shared/test_oopStorage_parperf.cpp:
```diff
@@ -152,7 +152,7 @@ class OopStorageParIterPerf::Task : public WorkerTask {
 class OopStorageParIterPerf::Closure : public OopClosure {
 public:
-  virtual void do_oop(oop* p) { guarantee(*p == NULL, "expected NULL"); }
+  virtual void do_oop(oop* p) { guarantee(*p == nullptr, "expected nullptr"); }
```
s/expected nullptr/expected null/

kimbarrett commented on test/hotspot/gtest/gc/shared/test_collectedHeap.cpp:
```diff
@@ -32,8 +32,8 @@ TEST_VM(CollectedHeap, is_in) {
-  // Test that NULL is not in the heap.
-  ASSERT_FALSE(heap->is_in(NULL)) << "NULL is unexpectedly in the heap";
+  // Test that nullptr is not in the heap.
+  ASSERT_FALSE(heap->is_in(nullptr)) << "nullptr is unexpectedly in the heap";
```
s/"nullptr is/"null is/

kimbarrett commented on test/hotspot/gtest/memory/test_guardedMemory.cpp:
```diff
@@ -30,7 +30,7 @@
 
 static void guarded_memory_test_check(void* p, size_t sz, void* tag) {
-  ASSERT_TRUE(p != NULL) << "NULL pointer given to check";
+  ASSERT_TRUE(p != nullptr) << "nullptr pointer given to check";
```
s/nullptr pointer/null pointer/

kimbarrett commented on test/hotspot/gtest/nmt/test_nmtpreinit.cpp:
```diff
@@ -84,11 +84,11 @@ class TestAllocations {
     p2 = os_realloc(os_malloc(10), 20);  // realloc, growing
     p3 = os_realloc(os_malloc(20), 10);  // realloc, shrinking
-    p4 = os_realloc(NULL, 10);           // realloc with NULL pointer
+    p4 = os_realloc(nullptr, 10);           // realloc with nullptr pointer
```
s/nullptr pointer/null pointer/

kimbarrett commented on test/hotspot/gtest/runtime/test_ThreadsListHandle.cpp:
```diff
@@ -213,18 +213,18 @@ TEST_VM(ThreadsListHandle, sanity) {
-    // Verify the current thread's hazard ptr is NULL:
-    EXPECT_EQ(ThreadsListHandleTest::get_Thread_threads_hazard_ptr(thr), (ThreadsList*)NULL)
-        << "thr->_threads_hazard_ptr must be NULL";
+    // Verify the current thread's hazard ptr is nullptr:
```
s/nullptr/null/

kimbarrett commented on test/hotspot/gtest/runtime/test_ThreadsListHandle.cpp:
```diff
@@ -397,9 +397,9 @@ TEST_VM(ThreadsListHandle, sanity) {
-      // Verify the current thread's hazard ptr is NULL:
-      EXPECT_EQ(ThreadsListHandleTest::get_Thread_threads_hazard_ptr(thr), (ThreadsList*)NULL)
-          << "thr->_threads_hazard_ptr must be NULL";
+      // Verify the current thread's hazard ptr is nullptr:
```
s/nullptr/null/

kimbarrett commented on test/hotspot/gtest/runtime/test_ThreadsListHandle.cpp:
```diff
@@ -445,18 +445,18 @@ TEST_VM(ThreadsListHandle, sanity) {
-    // Verify the current thread's hazard ptr is NULL:
-    EXPECT_EQ(ThreadsListHandleTest::get_Thread_threads_hazard_ptr(thr), (ThreadsList*)NULL)
-        << "thr->_threads_hazard_ptr must be NULL";
+    // Verify the current thread's hazard ptr is nullptr:
```
s/nullptr/null/

kimbarrett commented on test/hotspot/gtest/runtime/test_ThreadsListHandle.cpp:
```diff
@@ -565,18 +565,18 @@ TEST_VM(ThreadsListHandle, sanity) {
-    // Verify the current thread's hazard ptr is NULL:
-    EXPECT_EQ(ThreadsListHandleTest::get_Thread_threads_hazard_ptr(thr), (ThreadsList*)NULL)
-        << "thr->_threads_hazard_ptr must be NULL";
+    // Verify the current thread's hazard ptr is nullptr:
```
s/nullptr/null/

kimbarrett commented on test/hotspot/gtest/runtime/test_ThreadsListHandle.cpp:
```diff
@@ -643,18 +643,18 @@ TEST_VM(ThreadsListHandle, sanity) {
-    // Verify the current thread's hazard ptr is NULL:
-    EXPECT_EQ(ThreadsListHandleTest::get_Thread_threads_hazard_ptr(thr), (ThreadsList*)NULL)
-        << "thr->_threads_hazard_ptr must be NULL";
+    // Verify the current thread's hazard ptr is nullptr:
```
s/nullptr/null/

kimbarrett commented on test/hotspot/gtest/runtime/test_classLoader.cpp:
```diff
@@ -30,32 +30,32 @@
-  ASSERT_TRUE(bad_class_name) << "Function did not set bad_class_name with NULL class name";
-  ASSERT_TRUE(retval == NULL) << "Wrong package for NULL class name pointer";
+  TempNewSymbol retval = ClassLoader::package_from_class_name(nullptr, &bad_class_name);
+  ASSERT_TRUE(bad_class_name) << "Function did not set bad_class_name with nullptr class name";
```
s/nullptr/null/

kimbarrett commented on test/hotspot/gtest/runtime/test_classLoader.cpp:
```diff
@@ -30,32 +30,32 @@
-  ASSERT_TRUE(retval == NULL) << "Wrong package for NULL class name pointer";
+  TempNewSymbol retval = ClassLoader::package_from_class_name(nullptr, &bad_class_name);
+  ASSERT_TRUE(bad_class_name) << "Function did not set bad_class_name with nullptr class name";
+  ASSERT_TRUE(retval == nullptr) << "Wrong package for nullptr class name pointer";
```
second s/nullptr/null/

kimbarrett commented on test/hotspot/gtest/runtime/test_os_linux.cpp:
```diff
@@ -200,8 +200,8 @@ TEST_VM(os_linux, reserve_memory_special_huge_tlbfs_size_not_aligned_with_bad_re
       // as the area around req_addr contains already existing mappings, the API should always
-      // return NULL (as per contract, it cannot return another address)
-      EXPECT_TRUE(p == NULL) << " size = " << size
+      // return nullptr (as per contract, it cannot return another address)
```
s/nullptr/null/

coleenp replied:
This is technically correct.  Changing the comments like this seems really pointless.

kimbarrett commented on test/hotspot/gtest/runtime/test_os_linux_cgroups.cpp:
```diff
@@ -63,13 +63,13 @@ TEST(cgroupTest, set_cgroupv1_subsystem_path) {
   TestCase at_mount_root = {
     "/sys/fs/cgroup",       // mount_path
-    NULL,                   // root_path, ignored
+    nullptr,                   // root_path, ignored
```
comment mis-indented

kimbarrett commented on test/hotspot/gtest/runtime/test_os_linux_cgroups.cpp:
```diff
@@ -63,13 +63,13 @@ TEST(cgroupTest, set_cgroupv1_subsystem_path) {
   TestCase sub_path = {
     "/sys/fs/cgroup",       // mount_path
-    NULL,                   // root_path, ignored
+    nullptr,                   // root_path, ignored
```
comment mis-indented

kimbarrett commented on test/hotspot/gtest/testutils.hpp:
```diff
@@ -53,8 +53,8 @@ class GtestUtils : public AllStatic {
-// (ASSERT|EXPECT)_NOT_NULL: check that the given pointer is not NULL
-// (ASSERT|EXPECT)_NULL: check that the given pointer is NULL
+// (ASSERT|EXPECT)_NOT_NULL: check that the given pointer is not nullptr
+// (ASSERT|EXPECT)_NULL: check that the given pointer is nullptr
```
both s/nullptr/null/

kimbarrett reviewed (changes requested):
Mostly good.  A few nullptr -> null, and a couple other nits.

kimbarrett reviewed (approved):
Looks good.

dholmes-ora commented on test/hotspot/gtest/nmt/test_nmt_buffer_overflow_detection.cpp:
```diff
@@ -109,7 +109,7 @@ DEFINE_TEST(test_double_free, "header canary")
 
 static void test_invalid_block_address() {
-  // very low, like the result of an overflow or of accessing a NULL this pointer
+  // very low, like the result of an overflow or of accessing a nullptr this pointer
```
s/nullptr/null

This is referring to nullness as a concept.

dholmes-ora reviewed (approved):
Looks good.

Part of the cleanup with using `nullptr` is that we should never need to cast `nullptr` to another pointer type. I flagged a few cases below but certainly not all.

For text/comments the general rule is that when talking about the general concept of nullness then we can say "null", whereas if referring to an actual code artifact we use "nullptr". Sometimes either will be fine e.g "returns null on error", "returns nullptr on error".

Thanks.

kimbarrett commented:
> Part of the cleanup with using `nullptr` is that we should never need to cast `nullptr` to another pointer type. I flagged a few cases below but certainly not all.

I think this may eventually be true, but I don't think it is currently true.  I think we will need to add overloads for
std::nullptr_t in some places in order to remove some casts.  This is particularly true for templates, and I think all
of the places you mentioned are of this kind.

coleenp commented on test/hotspot/gtest/metaspace/metaspaceGtestCommon.hpp:
```diff
@@ -145,12 +145,12 @@ void mark_range(MetaWord* p, size_t word_size);
 
 //////////////////////////////////////////////////////////
-// Some helpers to avoid typing out those annoying casts for NULL
+// Some helpers to avoid typing out those annoying casts for nullptr
```
Like I said in the description, we can change the comments on a case-by-case basis in order to get the bulk of the change done mechanically.

coleenp replied:
ie, we should file a separate RFE for comments.

coleenp replied:
And in this comment nullptr makes perfect sense.

dholmes-ora replied:
Yeah I wasn't commenting on the use of nullptr in the comment :)

coleenp commented:
Thanks Julian, Kim and David for the reviews.  I fixed the null casts that you pointed out and the one comment, and fixed my copyright script (so updated the copyrights).  I'll wait for GHA to integrate.
